File: 09_RAC/09_RAC/ViewController.m
//
//  ViewController.m
//  09_RAC
//
//  Created by GeekerHua on 15/12/30.
//  Copyright © 2015年 GeekerHua. All rights reserved.
//

#import "ViewController.h"
#import "Person.h"
#import <ReactiveCocoa/ReactiveCocoa.h>
#import <ReactiveCocoa/RACEXTScope.h>  // 范围
#import "NetworkTools.h"

@interface ViewController ()
@property (strong, nonatomic) Person *person;    /**<   */
@property (weak, nonatomic) IBOutlet UILabel *nameLabel;
@property (weak, nonatomic) IBOutlet UIButton *demoButton;
@property (weak, nonatomic) IBOutlet UITextField *textField;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.person = [[Person alloc] init];
//    [self targetDemo];
    [[[NetworkTools sharedTools] request:GET URLString:@"http://www.weather.com.cn/data/sk/101010100.html" parameters:nil] subscribeNext:^(id x) {
        NSLog(@"%@",x);
    } error:^(NSError *error) {
        NSLog(@"%@",error);
    }];
}

- (void)notificationDemo{
    // takeUntile 表示信号有效期
    [[[[NSNotificationCenter defaultCenter] rac_addObserverForName:UIApplicationDidEnterBackgroundNotification object:nil] takeUntil:[self rac_willDeallocSignal]] subscribeNext:^(id x) {
        NSLog(@"%@",x);
    }];
    
}

- (void)textFieldDemo{
    
    [[self.textField rac_textSignal] subscribeNext:^(id x) {
        NSLog(@"%@",x);
    }];
}

/*
 1. 导入头文件 #import <ReactiveCocoa/RACEXTScope.h>
#import "NetworkTools.h"
 2. 外部写 @weakify(self)
 3. 内部写 @strongify(self)
 RAC的Block中只要出现self,一定会强引用
 */
- (void)targetDemo{
    @weakify(self)
    [[self.demoButton rac_signalForControlEvents:UIControlEventTouchUpInside] subscribeNext:^(id x) {
        @strongify(self)
        NSLog(@"%@",x);
        self.textField.text = @"hello";
    }];
    
}

- (void)targetDemo2{
    // OC解决循环引用
    __weak typeof(self) weakSelf = self;
    [[self.demoButton rac_signalForControlEvents:UIControlEventTouchUpInside] subscribeNext:^(id x) {
        // 为了避免block被传出控制器,且当前控制器销毁了,为了不影响block执行,使用strongSelf修饰
        __strong typeof(weakSelf) strongSelf = weakSelf;
        NSLog(@"%@",x);
        strongSelf.textField.text = @"hello";
    }];
    
}

- (void)kvoDemo{
    // 一旦收到next消息,在next的Block中执行相应处理,不需要移除监听者
    [RACObserve(self.person, name) subscribeNext:^(id x) {
        NSLog(@"%@", x);
        self.nameLabel.text = x;
    }];
    
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event{
    self.person.name = [NSString stringWithFormat:@"zhangsan %05d", arc4random_uniform(100000)];
}


- (void)dealloc
{
    NSLog(@"dealloc - %s",__FUNCTION__);
}
@end

File: 06-闭包的循环引用/06-闭包的循环引用/ViewController.swift
//
//  ViewController.swift
//  06-闭包的循环引用
//
//  Created by GeekerHua on 15/12/30.
//  Copyright © 2015年 GeekerHua. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    var finishedCallBack: ((msg: String) ->())?
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // 写法1 [weak self]  不对self进行强引用,可以随时释放
        loadData { [weak self] (msg) -> () in
            // 闭包是准备的代码,要写self
            print(msg + "\(self?.view)")
        }
    
//        // 写法2 [unowned self]  会记录self的地址,但不会强引用,风险,self被释放了,就会崩溃
//        loadData { [unowned self] (msg) -> () in
//            // 闭包是准备的代码,要写self
//            print(msg + "\(self.view)")
//        }
    }
    
    func demoOC() {
        //        OC 的写法,解决循环引用
        weak var weakSelf = self
        loadData { (msg) -> () in
            // 闭包是准备的代码,要写self
            print(msg + "\(weakSelf?.view)")
        }
    }

    func loadData(finished: (msg:String) -> ()) {
        dispatch_async(dispatch_get_global_queue(0, 0)) { () -> Void in
            print("耗时操作")
            dispatch_async(dispatch_get_main_queue(), { () -> Void in
//                finished(msg: "hello")
                self.finishedCallBack = finished
                self.working()
            })
        }
    }
    
    func working(){
        // 闭包不存在就甭
//        finishedCallBack!(msg:"hello")
        // 闭包不存在就不执行
        finishedCallBack?(msg:"hello")
    }
    
    deinit{
        print("释放了")
    }
}

